src/bit_machine/frame.rs

```diff
@@ -15,227 +15,283 @@
-        unsafe {
-            let p = self.data.offset((self.abs_pos + n) / 8);
-            *p & (1 << (7 - (self.abs_pos + n) % 8)) != 0
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u16(&mut self, value: u16, data: &mut [u8]) {
+        for idx in 0..16 {
+            self.write_bit(value & (1 << (15 - idx)) != 0, data);
         }
     }
 
-    pub(in crate::bit_machine) fn peek_bit(&self) -> bool {
-        unsafe {
-            let p = self.data.offset(self.abs_pos / 8);
-            *p & (1 << (7 - self.abs_pos % 8)) != 0
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u32(&mut self, value: u32, data: &mut [u8]) {
+        for idx in 0..32 {
+            self.write_bit(value & (1 << (31 - idx)) != 0, data);
         }
     }
 
-    pub(crate) fn read_bit(&mut self) -> bool {
-        unsafe {
-            let p = self.data.offset(self.abs_pos / 8);
-            let ret = *p & (1 << (7 - self.abs_pos % 8)) != 0;
-            self.abs_pos += 1;
-            ret
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u64(&mut self, value: u64, data: &mut [u8]) {
+        for idx in 0..64 {
+            self.write_bit(value & (1 << (63 - idx)) != 0, data);
         }
     }
 
-    READ_UNSIGNED!(read_u8, u8);
-    READ_UNSIGNED!(read_u16, u16);
-    READ_UNSIGNED!(read_u32, u32);
-    READ_UNSIGNED!(read_u64, u64);
+    /// Move the cursor forward by the given length.
+    pub(crate) fn move_cursor_forward(&mut self, len: usize) {
+        self.cursor += len;
+    }
 
-    /// Write a big-endian u64 value to the current write frame
-    pub(in crate::bit_machine) fn write_u64(&mut self, data: u64) {
-        for idx in 0..64 {
-            self.write_bit(data & (1 << (63 - idx)) != 0);
-        }
+    /// Move the cursor backward by the given length.
+    pub(crate) fn move_cursor_backward(&mut self, len: usize) {
+        self.cursor -= len;
     }
 
-    /// Write a big-endian u32 value to the current write frame
-    pub(in crate::bit_machine) fn write_u32(&mut self, data: u32) {
-        for idx in 0..32 {
-            self.write_bit(data & (1 << (31 - idx)) != 0);
+    /// Copy a bit string of given length from another frame into the present one.
+    pub(crate) fn copy_from(&mut self, other: &Self, len: usize, data: &mut [u8]) {
+        for i in 0..len {
+            let (other_byte_index, other_bit_index) = get_indices(other.cursor + i);
+            let bit = data[other_byte_index] & (1 << (7 - other_bit_index)) != 0;
+            self.write_bit(bit, data);
         }
     }
 
-    /// Write a big-endian u16 value to the current write frame
-    pub(in crate::bit_machine) fn write_u16(&mut self, data: u16) {
-        for idx in 0..16 {
-            self.write_bit(data & (1 << (15 - idx)) != 0);
+    /// Adds
+    pub fn with_data<'a>(&self, data: &'a [u8]) -> FrameData<'a> {
```

Comment:
nit: This is a constructor of FrameData inside `impl Frame`. Perhaps move to this `impl FrameData` ? Or call this function `to_frame_data` ?